components/Buttons/src/MDCFloatingButton.m

```diff
@@ -72,67 +105,406 @@ - (instancetype)initWithCoder:(NSCoder *)aDecoder {
     case MDCFloatingButtonShapeMini:
-      return CGSizeMake([[self class] miniDimension], [[self class] miniDimension]);
-    case MDCFloatingButtonShapeLargeIcon:
-      return CGSizeMake([[self class] defaultDimension], [[self class] defaultDimension]);
+      return CGSizeMake(MDCFloatingButtonMiniDimension, MDCFloatingButtonMiniDimension);
   }
 }
 
+- (CGSize)intrinsicContentSizeForModeExpanded {
+  const CGSize intrinsicTitleSize =
+      [self.titleLabel sizeThatFits:CGSizeMake(CGFLOAT_MAX, CGFLOAT_MAX)];
+  const CGSize intrinsicImageSize =
+      [self.imageView sizeThatFits:CGSizeMake(CGFLOAT_MAX, CGFLOAT_MAX)];
+  CGFloat intrinsicWidth = intrinsicTitleSize.width + intrinsicImageSize.width +
+      self.imageTitleSpacing + internalLayoutSpacingInsets.left +
+      internalLayoutSpacingInsets.right + self.contentEdgeInsets.left +
+      self.contentEdgeInsets.right;
+  CGFloat intrinsicHeight = MAX(intrinsicTitleSize.height, intrinsicImageSize.height) +
+      self.contentEdgeInsets.top + self.contentEdgeInsets.bottom + internalLayoutSpacingInsets.top +
+      internalLayoutSpacingInsets.bottom;
+  return CGSizeMake(intrinsicWidth, intrinsicHeight);
+}
+
 - (CGSize)intrinsicContentSize {
-  switch (_shape) {
-    case MDCFloatingButtonShapeDefault:
-      return CGSizeMake([[self class] defaultDimension], [[self class] defaultDimension]);
-    case MDCFloatingButtonShapeMini:
-      return CGSizeMake([[self class] miniDimension], [[self class] miniDimension]);
-    case MDCFloatingButtonShapeLargeIcon:
-      return CGSizeMake([[self class] defaultDimension], [[self class] defaultDimension]);
+  CGSize contentSize = CGSizeZero;
+  if (self.mode == MDCFloatingButtonModeNormal) {
+    contentSize = [self intrinsicContentSizeForModeNormal];
+  } else if (self.mode == MDCFloatingButtonModeExpanded) {
+    contentSize = [self intrinsicContentSizeForModeExpanded];
+  }
+
+  if (self.minimumSize.height > 0) {
+    contentSize.height = MAX(self.minimumSize.height, contentSize.height);
+  }
+  if (self.maximumSize.height > 0) {
+    contentSize.height = MIN(self.maximumSize.height, contentSize.height);
+  }
+  if (self.minimumSize.width > 0) {
+    contentSize.width = MAX(self.minimumSize.width, contentSize.width);
+  }
+  if (self.maximumSize.width > 0) {
+    contentSize.width = MIN(self.maximumSize.width, contentSize.width);
   }
+  return contentSize;
 }
 
 - (void)layoutSubviews {
   // We have to set cornerRadius before laying out subviews so that the boundingPath is correct.
   self.layer.cornerRadius = CGRectGetHeight(self.bounds) / 2;
   [super layoutSubviews];
+
+   if (self.mode == MDCFloatingButtonModeNormal) {
+    return;
+  }
+
+  // Position the imageView and titleView
+  //
+  // +------------------------------------+
+  // |    |  |  |  CEI TOP            |   |
+  // |CEI +--+  |       +-----+       |CEI|
+  // | LT ||SP|-- A --|Title|-- A --|RGT|
```

Comment:
ascii art needs a tweek. Other lines need to be longer after the "SP"

Reply:
I believe it's a GH font issue. Looks good in Xcode. 
![xcode-art](https://user-images.githubusercontent.com/1753199/33562171-9466abe6-d8e2-11e7-857d-d7ff3c6906e4.png)

Reply:
sgtm